File: SwiftOSXProofOfConcept/data2yaml/data/maygate.lua
sptbl["maygate"] = {

    files = {
        module = "maygate.c",
        header = "maygate.h",
        example = "ex_maygate.c",
    },

    func = {
        create = "sp_maygate_create",
        destroy = "sp_maygate_destroy",
        init = "sp_maygate_init",
        compute = "sp_maygate_compute",
    },

    params = {
        optional = {
            {
                name = "prob",
                type = "SPFLOAT",
                description = "Probability of maygate. This is a value between 0-1. The closer to 1, the more likely the maygate will let a signal through.",
                default = 0
            },

            {
                name = "mode",
                type = "int",
                description = "If mode is nonzero, maygate will output one sample triggers instead of a gate signal.",
                default = 0
            },
        }
    },

    modtype = "module",

    description = [[Maygate is a "maybe gate". It takes in a trigger, and then it will randomly decide to turn the gate on or not. One particular application for maygate is to arbitrarily turn on/off sends to effects. One specific example of this could be a randomized reverb throw on a snare.]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "trig",
            description = "This expects a trigger signal."
        }
    },

    outputs = {
        {
            name = "out",
            description = "Signal out."
        }
    }

}

File: AudioKit/Core/Soundpipe/modules/maygate.c
#include <stdlib.h>
#include "soundpipe.h"

int sp_maygate_create(sp_maygate **p)
{
    *p = malloc(sizeof(sp_maygate));
    return SP_OK;
}

int sp_maygate_destroy(sp_maygate **p)
{
    free(*p);
    return SP_OK;
}

int sp_maygate_init(sp_data *sp, sp_maygate *p)
{
    p->prob = 0.0;
    p->gate = 0;
    p->mode = 0;
    return SP_OK;
}

int sp_maygate_compute(sp_data *sp, sp_maygate *p, SPFLOAT *in, SPFLOAT *out)
{
    if(*in == 0) {
        if(p->mode) {
            *out = 0;
        } else {
            *out = p->gate;
        }
        return SP_OK;
    }

    if((1.0 * sp_rand(sp) / SP_RANDMAX) <= p->prob) {
        *out = 1;
        p->gate = 1;
    } else {
        *out = 0;
        p->gate = 0;
    }
    return SP_OK;
}
